File: WebMap.Plugin/Models/DynamicMarker.cs
using System.Collections.Generic;
using System.Numerics;

namespace WebMap.Models;

public sealed record DynamicMarker
{
    public required uint ObjectiveId { get; init; }
    public required string Label { get; init; }
    public required List<DynamicMarkerData> Data { get; init; }
    public required ushort RecommendedLevel { get; init; }
    public required bool ShouldRender { get; init; }
}

public abstract record MarkerData
{
    public uint Icon { get; init; }
    public required string Tooltip { get; init; }
    public required uint LevelId { get; init; }
    public required Vector3 Position { get; init; }
}

public sealed record DynamicMarkerData : MarkerData
{
    public required uint ObjectiveId { get; init; }
    public required float Radius { get; init; }
    public required uint MapId { get; init; }
    public required uint DataId { get; init; }
}

public sealed record QuestLinkMarker : MarkerData
{
    public required ushort RecommendedLevel { get; init; }
    
    public required uint SourceMapId { get; init; }
    public required uint TargetMapId { get; init; }
}

// TODO: Move somewhere else

public sealed record FlagMarker
{
    // public required uint MapId { get; init; }
}

File: WebMap.Web/src/net/mappers/dynamic-marker.ts
import {
    DynamicMarkerList as FBDynamicMarkerList,
    DynamicMarker as FBDynamicMarker,
    DynamicMarkerData as FBDynamicMarkerData,
} from '@/net/models'
import type { DynamicMarker, DynamicMarkerData, MarkerType } from '@/types/dynamic-marker'
import mapVector3 from './vector3'

function mapDynamicMarkerData(fb: FBDynamicMarkerData): DynamicMarkerData {
    const position = fb.position()
    return {
        levelId: fb.levelId(),
        objectiveId: fb.objectiveId(),
        tooltip: fb.tooltip()!,
        icon: fb.icon(),
        position: mapVector3(position!),
        radius: fb.radius(),
        mapId: fb.mapId(),
        dataId: fb.dataId(),
    }
}

export function mapDynamicMarker(fb: FBDynamicMarker, type: MarkerType): DynamicMarker {
    const data: DynamicMarkerData[] = new Array(fb.dataLength())
    for (let i = 0; i < fb.dataLength(); i++) {
        const item = fb.data(i)
        if (item) {
            data[i] = mapDynamicMarkerData(item)
        }
    }

    return {
        objectiveId: fb.objectiveId(),
        label: fb.label()!,
        data,
        recommendedLevel: fb.recommendedLevel(),
        shouldRender: fb.shouldRender(),
        type,
    }
}

export default function mapDynamicMarkerList(fb: FBDynamicMarkerList, type: MarkerType): DynamicMarker[] {
    const markers: DynamicMarker[] = new Array(fb.markersLength())
    for (let i = 0; i < fb.markersLength(); i++) {
        const marker = fb.markers(i)
        if (marker) {
            markers[i] = mapDynamicMarker(marker, type)
        }
    }
    return markers
}
